src/main/java/net/kyori/adventure/text/minimessage/parser/TokenParser.java

```diff
@@ -68,9 +74,39 @@ public static ElementNode parse(
     final @NotNull String message,
     final boolean strict
   ) {
-    final List<Token> tokens = tokenize(message);
+    // first resolve placeholders...
+    final String actualMessage = resolvePlaceholders(message, string -> tagNameChecker.test(string, false), templateResolver);
```

Comment:
I'd kinda like to have this run on a mutable token stream rather than going back & forth between StringBuilders, but this is at least simpler